Exclude Stream, Optional, and additional collection types from ClassFanOutComplexityCheck

Check documentation: https://checkstyle.sourceforge.io/config_metrics.html#ClassFanOutComplexity

Checkstyle currently automatically excludes a number of classes from the class fan-out complexity check, per the default value of the "excludedClasses" property of ClassFanOutComplexity.  This includes a number of base collection classes (List, HashMap, TreeSet, etc.) which were added per issue #31 in PR [#69](../pull/69).

Java 8's stream-based classes feel like they are similarly "core" as the standard collection classes, and more functionally-written code may end up using these as much if not more than the base collection classes already excluded.  There are some other core collection classes that also seem similarly basic and very "core".

My proposal is that the list of excluded classes gets expanded to include these classes.  Here is a proposed list, based on personal experience with real-life projects and some quick looking through the Java API:

* `java.util.Collection`
* `java.util.EnumSet`
* `java.util.LinkedHashMap`
* `java.util.LinkedHashSet`
* `java.util.Optional`
* `java.util.OptionalDouble`
* `java.util.OptionalInt`
* `java.util.OptionalLong`
* `java.util.stream.DoubleStream`
* `java.util.stream.IntStream`
* `java.util.stream.LongStream`
* `java.util.stream.Stream`

```
$ javac Test.java

$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="ClassFanOutComplexity">
      <property name="max" value="5" />
    </module>
  </module>
</module>

$ cat Test.java
import java.util.*;
import java.util.stream.*;

public class Test {
    public void fanOutComplexityUsage() {
        // Collections
        Collection<String> collection = new LinkedHashSet<>(List.of("A", "B"));
        EnumSet<Character.UnicodeScript> enumSet =
                EnumSet.of(Character.UnicodeScript.COMMON, Character.UnicodeScript.LATIN);
        Map<String, List<Integer>> map = new LinkedHashMap<>();

        // Streams & optionals
        Stream<String> stream = Stream.of("A", "B");
        Optional<String> optional = stream.findAny();
        IntStream intStream = IntStream.range(0, 10);
        OptionalInt optionalInt = intStream.findAny();
        LongStream longStream = LongStream.range(0, 10);
        OptionalLong optionalLong = longStream.findAny();
        DoubleStream doubleStream = DoubleStream.iterate(1.1, d -> d * 2).limit(5);
        OptionalDouble optionalDouble = doubleStream.average();
    }
}

$ java -jar checkstyle-8.34-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:4:1: Class Fan-Out Complexity is 12 (max allowed is 5). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 1 errors.
```

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java
@@ -77,6 +77,10 @@ public abstract class AbstractClassCouplingCheck extends AbstractCheck {
             "Set", "HashSet", "SortedSet", "TreeSet",
             "Map", "HashMap", "SortedMap", "TreeMap",
             "Override", "Deprecated", "SafeVarargs", "SuppressWarnings", "FunctionalInterface",
+            "Collection", "EnumSet", "LinkedHashMap", "LinkedHashSet", "Optional",
+            "OptionalDouble", "OptionalInt", "OptionalLong",
+            // java.util.stream.*
+            "DoubleStream", "IntStream", "LongStream", "Stream",
         }).collect(Collectors.toSet()));
 
     /** Package names to ignore. */

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassDataAbstractionCouplingCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassDataAbstractionCouplingCheck.java
@@ -88,11 +88,12 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  * Property {@code excludedClasses} - Specify user-configured class names to ignore.
  * Type is {@code java.lang.String[]}.
  * Default value is {@code ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte,
- * Character, Class, Deprecated, Deque, Double, Exception, Float, FunctionalInterface,
- * HashMap, HashSet, IllegalArgumentException, IllegalStateException,
- * IndexOutOfBoundsException, Integer, LinkedList, List, Long, Map, NullPointerException,
- * Object, Override, Queue, RuntimeException, SafeVarargs, SecurityException, Set, Short,
- * SortedMap, SortedSet, String, StringBuffer, StringBuilder, SuppressWarnings, Throwable,
+ * Character, Class, Collection, Deprecated, Deque, Double, DoubleStream, EnumSet, Exception,
+ * Float, FunctionalInterface, HashMap, HashSet, IllegalArgumentException, IllegalStateException,
+ * IndexOutOfBoundsException, IntStream, Integer, LinkedHashMap, LinkedHashSet, LinkedList, List,
+ * Long, LongStream, Map, NullPointerException, Object, Optional, OptionalDouble, OptionalInt,
+ * OptionalLong, Override, Queue, RuntimeException, SafeVarargs, SecurityException, Set, Short,
+ * SortedMap, SortedSet, Stream, String, StringBuffer, StringBuilder, SuppressWarnings, Throwable,
  * TreeMap, TreeSet, UnsupportedOperationException, Void, boolean, byte, char, double,
  * float, int, long, short, var, void}.
  * </li>

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheck.java
@@ -55,11 +55,12 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  * Property {@code excludedClasses} - Specify user-configured class names to ignore.
  * Type is {@code java.lang.String[]}.
  * Default value is {@code ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte,
- * Character, Class, Deprecated, Deque, Double, Exception, Float, FunctionalInterface,
- * HashMap, HashSet, IllegalArgumentException, IllegalStateException,
- * IndexOutOfBoundsException, Integer, LinkedList, List, Long, Map, NullPointerException,
- * Object, Override, Queue, RuntimeException, SafeVarargs, SecurityException, Set, Short,
- * SortedMap, SortedSet, String, StringBuffer, StringBuilder, SuppressWarnings, Throwable,
+ * Character, Class, Collection, Deprecated, Deque, Double, DoubleStream, EnumSet, Exception,
+ * Float, FunctionalInterface, HashMap, HashSet, IllegalArgumentException, IllegalStateException,
+ * IndexOutOfBoundsException, IntStream, Integer, LinkedHashMap, LinkedHashSet, LinkedList, List,
+ * Long, LongStream, Map, NullPointerException, Object, Optional, OptionalDouble, OptionalInt,
+ * OptionalLong, Override, Queue, RuntimeException, SafeVarargs, SecurityException, Set, Short,
+ * SortedMap, SortedSet, Stream, String, StringBuffer, StringBuilder, SuppressWarnings, Throwable,
  * TreeMap, TreeSet, UnsupportedOperationException, Void, boolean, byte, char, double,
  * float, int, long, short, var, void}.
  * </li>

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java
@@ -57,8 +57,8 @@ public class ClassFanOutComplexityCheckTest extends AbstractModuleTestSupport {
         checkConfig.addAttribute("max", "0");
 
         final String[] expected = {
-            "6:1: " + getCheckMessage(MSG_KEY, 3, 0),
-            "38:1: " + getCheckMessage(MSG_KEY, 1, 0),
+            "7:1: " + getCheckMessage(MSG_KEY, 3, 0),
+            "39:1: " + getCheckMessage(MSG_KEY, 1, 0),
         };
 
         verify(checkConfig, getPath("InputClassFanOutComplexity.java"), expected);
@@ -202,8 +202,8 @@ public class ClassFanOutComplexityCheckTest extends AbstractModuleTestSupport {
         checkConfig.addAttribute("excludeClassesRegexps", "^Inner.*");
 
         final String[] expected = {
-            "6:1: " + getCheckMessage(MSG_KEY, 2, 0),
-            "38:1: " + getCheckMessage(MSG_KEY, 1, 0),
+            "7:1: " + getCheckMessage(MSG_KEY, 2, 0),
+            "39:1: " + getCheckMessage(MSG_KEY, 1, 0),
         };
 
         verify(checkConfig, getPath("InputClassFanOutComplexity.java"), expected);
@@ -218,8 +218,8 @@ public class ClassFanOutComplexityCheckTest extends AbstractModuleTestSupport {
         checkConfig.addAttribute("excludeClassesRegexps", "");
 
         final String[] expected = {
-            "6:1: " + getCheckMessage(MSG_KEY, 3, 0),
-            "38:1: " + getCheckMessage(MSG_KEY, 1, 0),
+            "7:1: " + getCheckMessage(MSG_KEY, 3, 0),
+            "39:1: " + getCheckMessage(MSG_KEY, 1, 0),
         };
 
         verify(checkConfig, getPath("InputClassFanOutComplexity.java"), expected);

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexity.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexity.java
@@ -2,6 +2,7 @@ package com.puppycrawl.tools.checkstyle.checks.metrics.classfanoutcomplexity;
 
 import javax.naming.*;
 import java.util.*;
+import java.util.stream.*;
 
 public class InputClassFanOutComplexity {
     private class InnerClass { //singleline comment
@@ -46,4 +47,25 @@ class InputMultiDimensionalArray {
     public  Object[][] get() {
         return new Object[][]{};
     }
+}
+
+class InputCollectionsExt {
+    private Collection col; // ok
+    private EnumSet enumSet; // ok
+    private LinkedHashMap map; // ok
+    private LinkedHashSet set; // ok
+}
+
+class InputOptionals {
+    private Optional<Long> op1; // ok
+    private OptionalInt op2; // ok
+    private OptionalLong op3; // ok
+    private OptionalDouble op4; // ok
+}
+
+class InputStreams {
+    private Stream s1; // ok
+    private IntStream s2; // ok
+    private LongStream s3; // ok
+    private DoubleStream s4; // ok
 }

--- a/src/xdocs/config_metrics.xml
+++ b/src/xdocs/config_metrics.xml
@@ -302,15 +302,16 @@ public class Test
               <td>Specify user-configured class names to ignore.</td>
               <td><a href="property_types.html#String.5B.5D">String[]</a></td>
               <td><code>
-                ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte, Character,
-                Class, Deprecated, Deque, Double, Exception, Float, FunctionalInterface,
-                HashMap, HashSet, IllegalArgumentException, IllegalStateException,
-                IndexOutOfBoundsException, Integer, LinkedList, List, Long, Map,
-                NullPointerException, Object, Override, Queue, RuntimeException,
-                SafeVarargs, SecurityException, Set, Short, SortedMap, SortedSet,
-                String, StringBuffer, StringBuilder, SuppressWarnings, Throwable,
-                TreeMap, TreeSet, UnsupportedOperationException, Void, boolean,
-                byte, char, double, float, int, long, short, var, void</code>
+                ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte,
+                Character, Class, Collection, Deprecated, Deque, Double, DoubleStream, EnumSet,
+                Exception, Float, FunctionalInterface, HashMap, HashSet, IllegalArgumentException,
+                IllegalStateException, IndexOutOfBoundsException, IntStream, Integer,
+                LinkedHashMap, LinkedHashSet, LinkedList, List, Long, LongStream, Map,
+                NullPointerException, Object, Optional, OptionalDouble, OptionalInt, OptionalLong,
+                Override, Queue, RuntimeException, SafeVarargs, SecurityException, Set, Short,
+                SortedMap, SortedSet, Stream, String, StringBuffer, StringBuilder,
+                SuppressWarnings, Throwable, TreeMap, TreeSet, UnsupportedOperationException, Void,
+                boolean, byte, char, double, float, int, long, short, var, void</code>
               </td>
               <td>5.7</td>
             </tr>
@@ -685,15 +686,16 @@ class Foo {
               <td>Specify user-configured class names to ignore.</td>
               <td><a href="property_types.html#String.5B.5D">String[]</a></td>
               <td><code>
-                ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte, Character,
-                Class, Deprecated, Deque, Double, Exception, Float, FunctionalInterface,
-                HashMap, HashSet, IllegalArgumentException, IllegalStateException,
-                IndexOutOfBoundsException, Integer, LinkedList, List, Long, Map,
-                NullPointerException, Object, Override, Queue, RuntimeException,
-                SafeVarargs, SecurityException, Set, Short, SortedMap, SortedSet,
-                String, StringBuffer, StringBuilder, SuppressWarnings, Throwable,
-                TreeMap, TreeSet, UnsupportedOperationException, Void, boolean,
-                byte, char, double, float, int, long, short, var, void</code>
+                ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte,
+                Character, Class, Collection, Deprecated, Deque, Double, DoubleStream, EnumSet,
+                Exception, Float, FunctionalInterface, HashMap, HashSet, IllegalArgumentException,
+                IllegalStateException, IndexOutOfBoundsException, IntStream, Integer,
+                LinkedHashMap, LinkedHashSet, LinkedList, List, Long, LongStream, Map,
+                NullPointerException, Object, Optional, OptionalDouble, OptionalInt, OptionalLong,
+                Override, Queue, RuntimeException, SafeVarargs, SecurityException, Set, Short,
+                SortedMap, SortedSet, Stream, String, StringBuffer, StringBuilder,
+                SuppressWarnings, Throwable, TreeMap, TreeSet, UnsupportedOperationException, Void,
+                boolean, byte, char, double, float, int, long, short, var, void</code>
               </td>
               <td>5.7</td>
             </tr>

